### ISSUE TITLE
Sound Not Precise Enough to Teach Much with Music

### ISSUE THREAD

**Issue by maryfries** [jmoenig/Snap#1807]
We are writing a sound project into Unit 3 of BJC, and I'm finding the effort limited by Snap's sound functionality. Here are two examples:

![image](https://user-images.githubusercontent.com/6549787/28282002-e3cd0bee-6af6-11e7-9aec-32f9d8545768.png)
Why is this not playing in a **round**?
https://snap.berkeley.edu/snapsource/snap.html#present:Username=mfries&ProjectName=Bug%3A%20sound%20round%20example

Also, this project used to play **chords**, but now it sounds terrible. What happened?
http://snap.berkeley.edu/snapsource/snap.html#present:Username=mfries&ProjectName=U3%20play%20chord

Launching sound blocks used to work better than this... Can we get the older functionality back?

**Comment by jmoenig** [comment]
Harmonic music is all about being in synch, i.e. getting the timing to be super precise. for this we need Snap's clock to be faster than what HTML's ```getAnimationFrame()``` feature is providing. In earlier versions Snap was clocked much faster internally, but then browsers started to no longer support intervals that are faster than 50-60 fps. There is a trick in Snap!, however, that really makes synching super precise again, it's playing harmonic or polyphonic music in turbo mode. This doesn't speed up the music, since those are "timed" interpolated operations, instad it makes them really precise.

So, in short: for polyphony it's best to use turbo - or you put mor effort into synchronizing measures with broadcast "synching" events.

The chords projects sounds great (and very nice!) to me. What's wrong with it?

**Comment by maryfries** [comment]
Turbo mode cleans up the timing, but it doesn't improve the clicking sound. I think that's what's messing up the chords project, which sounds terrible either with turbo mode on or off for me.

I think this is related to [#1015](https://github.com/jmoenig/Snap--Build-Your-Own-Blocks/issues/1015) 

Two interesting points on this:
1) It's way worse with my digital headphones. (But still bad without them.)
2) A colleague (who doesn't claim to have a musical ear) didn't notice the problem at all until I pointed it out. To my ear, it's really bad, but I have experience with music and sound editing, so I I suppose I listen for nuances like that. Together, we concluded that if a musically inclined kid hears this, they will likely conclude, "why would anyone use Snap to make music?" though perhaps a non-musical kid will hear no problems and get intrigued by the project. Still, I'm hesitant to highlight this weakness because the musically-inclined kids will hear it...

Is it possible to try tapering the ends of the sounds played by PLAY NOTE instead of just cutting them off? There is a click even with playing a single note, and I think they are stacking up.

**Comment by maryfries** [comment]
I double checked these sounds with a professional musician at EDC, and she agreed that this is highly problematic.

**Comment by micahcowan** [comment]
They sound problematic to me as well (also a musician).

I can also confirm that the "round" project's timing is off unless turbo is on.

The "clicks" in the chords project (and round) occur at the ends of notes (or in between), never at the start. I've frequently heard these clicks when waveforms are interrupted abruptly - when the waveform didn't complete, or jumps to a new waveform without fade. Specifically, when the signal output value jumps around. The usual cure to this is to use attack/release ramping (fade the sound in at the start, and out at the end, for a fraction of a second); however, for some waveforms, such as the sine waveform it sounds like _Snap!_ is using, you can cure it without attack or release, by ensuring that the sound always starts/stops at the start of the waveform (that is, when the output signal is zero). Probably, the reason clicking doesn't happen at the start of the first sound (at least, for me), is that the sound begins at the start of the waveform.

Other waveforms, like square wave (which by definition consists of nothing _but_ the signal value "jumping" between extremes), always sound like they click at the start and stop of the tone, unless there's an attack/release ramp.

@maryfries, is there a date you could provide where you are fairly confident that the sound was not producing these clicks? Or have the clicks always been present, and it's just the timing issue that is newer? If I have a "known good" time, it could help to look into the code to see what changed between now and then (and to compare the resulting sound), to try and find a solution.

Of course, if part of the problem is due to browser changes during that time, then looking at _Snap!_ history won't end up helping... still, if there's a way to introduce a mild attack/release, that should help considerably, regardless of whether it's unrelated to what introduced the clicking in the first place.

**Comment by maryfries** [comment]
Thanks for the confirmation! I agree that this is likely is the problem. (I have a whole pile of sticky notes with squiggles on them in my recycle bin today from trying to explain this issue.) ;)

I don't think it will be easily fixed by changing the waveform, though, because the user can change the note length and tempo. I think it could work, however, to just taper off the very end of the note played by PLAY NOTE. You can hear that click at the end of playing just a single note.

I'm not great at dates, but my "play chord" project (that once worked) is dated 9/25/15. Thanks!

**Comment by micahcowan** [comment]
Yeah, I also don't think changing the waveform would help. The sinewave is already the most gentle possible waveform, and still causes the issue.

Ensuring that the tone ends at appropriate points on the waveform would help, but that's too low-level of an operation to probably be a useful option for _Snap!_ - and anyway, it would _only_ help with the current sine waveform. If ever _Snap!_ added the option to select what waveform to play, it would stop working for those. Tapering off ("releasing") is the cheapest, and most universal, way to fix the issue.

Tapering on ("attack") doesn't seem to be necessary for the current waveform, but would be for some others (square at least), probably worth doing one if we're doing the other, anyway.

**Comment by maryfries** [comment]
I'm sorry. I wasn't clear. I also think we should taper the release. :)

**Comment by micahcowan** [comment]
Ignore that commit; the follow-on commit that replaces it is better in a few respects (including that quick notes will sound as they should). The pull request is #1810.

**Comment by micahcowan** [comment]
Some experimental work was done to try to implement attack and release in _Snap!_ (see the discussion in #1810), but it's looking like it introduces even more timing problems with the notes when it's used, at least under some circumstances. It seems to cause the browser to work too hard to keep the timing.

I don't think the clicks are new to _Snap!_, but have probably always been there, unlike perhaps the timing issues you noted, that require turbo mode to mitigate.

**Comment by micahcowan** [comment]
New fixes in #1810 seem to do the trick now!

**Comment by jmoenig** [comment]
Audio has been rewritten for Snap5, including a remedy for the "clicks". This can now be closed. Thanks!
